Add toIP to CiliumLocalRedirectPolicy redirectBackend

liyihuang opened the pull request:
see commit message

Fixes: https://github.com/cilium/cilium/issues/41671

```release-note
Add toIP to CiliumLocalRedirectPolicy redirectBackend so it works with latest GKE metadata server
```

liyihuang committed:
pkg/lrp: implement the toIP

In the GKE environment, GKE metadata-server starts
to listen on --addr=169.254.169.252:988. This commit
implements the overrideIP field so we can redirect
to right IP address.

Before this commit, if we use the following lrp policy

apiVersion: cilium.io/v2
kind: CiliumLocalRedirectPolicy
metadata:
  name: gke-metadata-server-redirect-80
  namespace: kube-system
spec:
  redirectBackend:
    localEndpointSelector:
      matchLabels:
        k8s-app: "gke-metadata-server"
    toPorts:
    - port: "988"
      protocol: "TCP"

  redirectFrontend:
    addressMatcher:
      ip: "169.254.169.254"
      toPorts:
      - port: "80"
        protocol: "TCP"
we get
root@gke-liyi-gke-test-liyi-np-0-3fa00d47-pzxn:/home/cilium# cilium
service list
ID   Frontend                 Service Type    Backend
18   169.254.169.254:80/TCP   LocalRedirect   1 => 172.16.64.7:988/TCP
(active)

172.16.64.7 is the node IP of the since gke-metadata-server is running
in
host network ns.

with the new field, we can get to the right IP address where server is
binding

root@gke-liyi-gke-test-liyi-np-0-3fa00d47-pzxn:/home/cilium# cilium
service list
ID   Frontend                 Service Type    Backend
19   169.254.169.254:80/TCP   LocalRedirect   1 =>
169.254.169.252:988/TCP (active)

Signed-off-by: Liyi Huang <liyi.huang@isovalent.com>

liyihuang committed:
pkg/loadbalancer: add test for LRP with toIP

Add a new test data file to verify the functionality
of the local redirect policy using the toIP field.

Signed-off-by: Liyi Huang <liyi.huang@isovalent.com>

liyihuang committed:
Documentation: Add section for LRP toIP

Explains the use of the `toIP` field and add the GKE example

Signed-off-by: Liyi Huang <liyi.huang@isovalent.com>

joamaki reviewed (approved):
Nice!

qmonnet commented on Documentation/network/kubernetes/local-redirect-policy.rst:
```diff
@@ -608,3 +612,37 @@ and exposed to user space in versions >= 5.12.
+-----------
+The local redirect policy supports overriding the IP address of the backend pod.
+This can be used to redirect traffic to a different IP address than the one assigned by Kubernetes.
+This is useful when the backend pod is running in the host network namespace where it could only listen on the specific address from loopback interface(such as GKE metadata server).
```
Nit: missing a space before the parenthesis

```suggestion
This is useful when the backend pod is running in the host network namespace where it could only
listen on the specific address from loopback interface (such as GKE metadata server).
```

qmonnet commented on Documentation/network/kubernetes/local-redirect-policy.rst:
```diff
@@ -608,3 +612,37 @@ and exposed to user space in versions >= 5.12.
+This can be used to redirect traffic to a different IP address than the one assigned by Kubernetes.
+This is useful when the backend pod is running in the host network namespace where it could only listen on the specific address from loopback interface(such as GKE metadata server).
+
+you can use the ``overrideIP`` field to override the IP address of the backend pod.
```
```suggestion
You can use the ``overrideIP`` field to override the IP address of the backend pod.
```

qmonnet commented on Documentation/network/kubernetes/local-redirect-policy.rst:
```diff
@@ -608,3 +612,37 @@ and exposed to user space in versions >= 5.12.
+
+.. note::
+
+    Even though you can configure the ``overrideIP`` field with ``redirectFrontend`` in the policy, it will not be used.
```
I find the phrasing could be improved. What will not be used in that case, `overrideIP` or `redirectFrontend`? Maybe simply:


```suggestion
    Field ``overrideIP`` has no effect when used for ``redirectFrontend``.
```

liyihuang replied:
I think yours is accurate.

qmonnet reviewed (changes requested):
Thanks! Here are some small doc nits from my side

borkmann commented on pkg/defaults/defaults_linux.go:
```diff
@@ -10,5 +10,5 @@ import (
 	// AddressScopeMax controls the maximum address scope for addresses to be
 	// considered local ones with HOST_ID in the ipcache
-	AddressScopeMax = int(netlink.SCOPE_LINK) - 1
+	AddressScopeMax = int(netlink.SCOPE_LINK) + 1
```
scope link + 1 == netlink.SCOPE_HOST, no?

Otherwise lgtm, I would split this commit off into a new PR, and mark it for backport.

liyihuang replied:
Sure. May I know why do you think we should backport it? I thought we will just configure it as 254 for the old version. Do you still think we should keep this as the hidden flag for the agent?

borkmann replied:
I mean we could do at least to 1.18.. given it fixes an issue for users with bpf host routing, and they might not be aware how to fix it. So if we can do this by default, it might be better.

borkmann replied:
> Do you still think we should keep this as the hidden flag for the agent?

If we change the default so that things work out of the gate, we could leave it for now as hidden.

liyihuang replied:
> I mean we could do at least to 1.18.. given it fixes an issue for users with bpf host routing, and they might not be aware how to fix it. So if we can do this by default, it might be better.

btw this commit itslef doesn't help the GKE meta data server for bpf host routing issue since it requires other commits in PR together. That's why I have it in this PR. It could help other use cases like HCP, I will seperate this commit as the seperate PR.

qmonnet reviewed (approved):
Doc update looks good, thank you!

sypakine reviewed (approved):
While I have expressed my thoughts about the `overrideIP` approach, I am not obstinate 😄 . If others are fine with the API design, we should move ahead with it.

squeed commented on pkg/k8s/apis/cilium.io/v2/clrp_types.go:
```diff
@@ -103,6 +104,16 @@ type PortInfo struct {
+	//
+	// +kubebuilder:validation:Pattern=`((^\s*((([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5]))\s*$)|(^\s*((([0-9A-Fa-f]{1,4}:){7}([0-9A-Fa-f]{1,4}|:))|(([0-9A-Fa-f]{1,4}:){6}(:[0-9A-Fa-f]{1,4}|((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3})|:))|(([0-9A-Fa-f]{1,4}:){5}(((:[0-9A-Fa-f]{1,4}){1,2})|:((25[0-5]|2[0-4]\d|1 ...
+	// +kubebuilder:validation:Optional
+	OverrideIP string `json:"overrideIP,omitempty"`
```
Move this to a property of RedirectBackend -- having it in the ToPorts field is confusing.

liyihuang replied:
@squeed I agree, but I'm not sure how we gonna support multiple IP for a pod. If we support a list of IP, how we can map the IP to the port if we move this to RedirectBackend?

squeed commented on pkg/k8s/apis/cilium.io/v2/clrp_types.go:
```diff
@@ -103,6 +104,16 @@ type PortInfo struct {
+	// other interfaces for which Cilium cannot obtain the IP from Kubernetes.
+	// This field is only valid for RedirectBackend.
+	//
+	// +kubebuilder:validation:Pattern=`((^\s*((([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5]))\s*$)|(^\s*((([0-9A-Fa-f]{1,4}:){7}([0-9A-Fa-f]{1,4}|:))|(([0-9A-Fa-f]{1,4}:){6}(:[0-9A-Fa-f]{1,4}|((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3})|:))|(([0-9A-Fa-f]{1,4}:){5}(((:[0-9A-Fa-f]{1,4}){1,2})|:((25[0-5]|2[0-4]\d|1 ...
```
it's 2025 :-) , we can now write
```
// +kubebuilder:validation:XValidation:rule="isIP(self)", message="overrideIP must be a valid IP address"
```

liyihuang replied:
will push another commit for this after figuring out  https://github.com/cilium/cilium/pull/41645#discussion_r2391489953

aditighag reviewed (changes requested):
Thanks for working on this.  

I agree that the `overrideIP` field should be defined under `RedirectBackend`. If there is a use case for multiple override IP addresses, this could be defined as a slice. 

> * I'm not sure if that gonna work well with https://docs.cilium.io/en/latest/network/kubernetes/local-redirect-policy/#servicematcher with dual stack or multi network features.
> * this doesn't work if there are mulitple ports in the frontend and pod listenning to different IP with ports in the backend. Users have to break one LRP polices to multiple polices where I think it's not aligned with the UX that we provide with the current API(we support users writing LRP with mulitple ports but have to ask them break into multiple when using the `RedirectBackend.ToAddress`).

We could add a validation so that the overrideIP field can only be used with AddressMatcher LRPs. 

> 
> I'm open to change the API to `RedirectBackend.ToAddress` if we all think that's better design and UX
> 
> @joamaki @aditighag since you worked on this field. Hope you can judge which is overall better design.

aditighag commented on Documentation/network/kubernetes/local-redirect-policy.rst:
```diff
@@ -608,3 +612,38 @@ and exposed to user space in versions >= 5.12.
+The local redirect policy supports overriding the IP address of the backend pod.
+This can be used to redirect traffic to a different IP address than the one assigned by Kubernetes.
+This is useful when the backend pod is running in the host network namespace where it could only 
+listen on the specific address from loopback interface (such as GKE metadata server).
```
Can you add an example with GKE use case under this section - https://docs.cilium.io/en/latest/network/kubernetes/local-redirect-policy/#use-cases?

liyihuang replied:
sure. I will add that later.

liyihuang committed:
ipkg/lrp: add toIP to RedirectBackend

In GKE environment, GKE meta server can be running on the host network
namespace as a DaemonSet and listening on the loopback secondary IP.
Cilium LRP just read the IP address from pod status and doesn't know the
loopback secondary IP. Adding toIP field to the RedirectBackend struct
and CRD to support this use case.

- Add the toIP field to the RdirectBackend struct and CRD.
- Update getSanitizedLocalRedirectPolicy to parse the field.

Signed-off-by: Liyi Huang <liyi.huang@isovalent.com>

qmonnet commented on Documentation/network/kubernetes/local-redirect-policy.rst:
```diff
@@ -521,9 +521,106 @@ will get directed to cluster DNS pods backed by the ``kube-dns`` service.
             11   10.96.0.10:53      LocalRedirect   1 => 10.244.1.49:53 (active)
 
+GKE Metadata Server with OverrideIP
+------------------------------------
```
```suggestion
GKE Metadata Server with OverrideIP
-----------------------------------
```

qmonnet commented on Documentation/network/kubernetes/local-redirect-policy.rst:
```diff
@@ -521,9 +521,106 @@ will get directed to cluster DNS pods backed by the ``kube-dns`` service.
+GKE Metadata Server with OverrideIP
+------------------------------------
+The GKE metadata server is a pod running in the host network namespace. The new version of GKE metadata server only
+listens a specific secondary address from the loopback interface. GKE uses the IP table rule to redirect the traffic
```
```suggestion
listens on a specific secondary address from the loopback interface. GKE uses the IP table rule to redirect the traffic
```

qmonnet commented on Documentation/network/kubernetes/local-redirect-policy.rst:
```diff
@@ -521,9 +521,106 @@ will get directed to cluster DNS pods backed by the ``kube-dns`` service.
+To optimize the traffic path, you can use the local redirect policy with ``overrideIP`` field to redirect the traffic to the metadata server 
+pod on the local node with Cilium eBPF host routing.
+
+Assume you have the following helm values for Cilium:
```
```suggestion
Assuming you have the following Helm values for Cilium:
```

qmonnet commented on Documentation/network/kubernetes/local-redirect-policy.rst:
```diff
@@ -521,9 +521,106 @@ will get directed to cluster DNS pods backed by the ``kube-dns`` service.
+   localRedirectPolicies:
+     enabled: true
+
+The GKE metadata server pod is deployed and bind to the ``169.254.169.252:988``.
```
```suggestion
The GKE metadata server pod is deployed and binds to ``169.254.169.252:988``.
```

qmonnet commented on Documentation/network/kubernetes/local-redirect-policy.rst:
```diff
@@ -521,9 +521,106 @@ will get directed to cluster DNS pods backed by the ``kube-dns`` service.
+     masquerade: true
+   kubeProxyReplacement: true
+   localRedirectPolicies:
+     enabled: true
```
```suggestion
.. code-block:: yaml

   bpf:
     masquerade: true
   kubeProxyReplacement: true
   localRedirectPolicies:
     enabled: true
```

qmonnet commented on Documentation/network/kubernetes/local-redirect-policy.rst:
```diff
@@ -521,9 +521,106 @@ will get directed to cluster DNS pods backed by the ``kube-dns`` service.
+            protocol: "TCP"
+
+After applying the local redirect policy, the traffic to the metadata server will be redirected to the node-local backend pod.
+You can check local redirect policy to see the policy has been applied from cilium agent through the following commands:
```
```suggestion
You can check local redirect policy to see the policy has been applied from the Cilium agent with the following command:
```

qmonnet commented on Documentation/network/kubernetes/local-redirect-policy.rst:
```diff
@@ -521,9 +521,106 @@ will get directed to cluster DNS pods backed by the ``kube-dns`` service.
+
+    $ cilium-dbg service list | grep LocalRedirect
+    19   169.254.169.254:80/TCP   LocalRedirect   1 => 169.254.169.252:988/TCP (active)
+
+
+
```
```suggestion

```

(one blank line is enough, you've got three)

qmonnet commented on Documentation/network/kubernetes/local-redirect-policy.rst:
```diff
@@ -521,9 +521,106 @@ will get directed to cluster DNS pods backed by the ``kube-dns`` service.
+
+You can use ``tcpdump`` on the host to verify the traffic is redirected to the node-local backend pod.
+
+As you can see from the following output the packet is sent to the node-local backend pod through ``lxc2140a2364964`` directly from the pod(10.36.1.28).
```
```suggestion
As you can see from the following output the packet is sent to the node-local backend pod through ``lxc2140a2364964`` directly from the pod (``10.36.1.28``).
```

qmonnet reviewed (changes requested):
Docs look good, I only have some minor style or grammar nits.

squeed commented on pkg/k8s/apis/cilium.io/v2/clrp_types.go:
```diff
@@ -147,6 +147,14 @@ type RedirectBackend struct {
 	// +kubebuilder:validation:Required
 	ToPorts []PortInfo `json:"toPorts"`
+
+	// OverrideIP is an IP address used to override the Pod IP.
```
Quick question: does "Pod IP" here mean that the source IP is changed, or the destination?

liyihuang replied:
the destination IP address.

The use case here is the dst pod is on host NS and could listen to some IP that's different from pod IP that we get from k8s.

You could read here for more details

https://github.com/cilium/cilium/issues/41671#issue-3414455168

joestringer replied:
Perhaps `ToIP` would be a bit more consistent with the `ToPorts` naming of the other override?

liyihuang replied:
thanks. that's a good point and `ToIP` sounds intuitive for users. The only thing is `ToPorts` is required field and `ToIP`  is totally optional and is only needed when people need to override it.

squeed replied:
Joe's suggestion is a good one, `toIP`'s behavior basically matches that of `toPorts`. I would document it a bit like this:

> ToIP overrides the destination backend's IP. Optional; when not set, the pod's IP is used. Use this if the backend pod's service is not available via the pod's published PodIPs.

liyihuang replied:
thanks. I will change it in my next commit for this

squeed commented on pkg/k8s/apis/cilium.io/v2/clrp_types.go:
```diff
@@ -147,6 +147,14 @@ type RedirectBackend struct {
+	// other interfaces for which Cilium cannot obtain the IP from Kubernetes.
+	// +kubebuilder:validation:XValidation:rule="isIP(self)", message="overrideIP must be a valid IP address"
+	// +kubebuilder:validation:Optional
+	OverrideIP string `json:"overrideIP,omitempty"`
```
How does this work with dual-stack? Does this need to be a slice?

liyihuang replied:
There were a lot of discussions back and forth and this one.

https://github.com/cilium/cilium/pull/41645#issuecomment-3563855329


In my view, there are 2 ways to solve this, and I'm open to impelment what makes the most sense to us.

1. just use the string and ask users to write 2 LRP for dual stack. that's how we do the unit test now. https://github.com/cilium/cilium/blob/98ae1b83407a5ac27bccb7ffbb304d6c52a122e5/pkg/loadbalancer/redirectpolicy/testdata/address.txtar#L21-L22.


2. use the slice and limit the length to be 2 max and only take 1 ipv4 and 1 ipv6 for length 2.(if we just take a slice with random length of IP, it will be impossible to know which one to replace for the orginal dst IP)


I actually implemented 2 and revert to 1 since I can see that's how we test dual stack in the unit test.


this can cover dual stack use case but can't handle multi NIC use case where I dont think it's common.

joestringer replied:
Putting a slice here means ambiguity about which backend the traffic will actually be redirected to. To resolve that we'd probably need to add further fields to determine IP selection within the slice. I think that's a bit too much scope creep for the problem we're looking to solve. I agree that the option 2 above seems a bit too restrictive.

squeed replied:
easy enough!

squeed replied:
As long as there's a way for this to be dual-stack, I'm satisfied.

liyihuang replied:
I just realized an issue with this method, and I'm not sure how much it will impact the users.

let's say users have the dual stack enviroment, they can use 2 address based LRP like what we did in the unit test to cover the dual stack.

when they have a service based LRP for dual stack, the current method will only override single stack and I guess the other stack for LRP doesn't work any more since we have some filtering down the line to match the frontend stack with backend stack ip familiy.

the workaround could be asking user to use address based LRP to match the service IP but we all know the cluster IP is not persistent

If we consider this field for GKE metadata server use case, I think that's enough. I just dont know what other use case could be that they have to use service based LRP and have to override the IP.


Maybe I'm thinking too much and we should just implement it to cover this use case and wait for the feedback.

@joestringer @squeed

joestringer replied:
Hmm. The naive way to think about this would be "match service X => redirect to backend <IP>" would NAT all traffic to that IP. It's technically possible to NAT64 or NAT46 this, so the simplest interpretation of the policy would expect address family translation as well. Except I presume we didn't implement that as it's more complicated. So even if we have just one IP address here, we're forcing the constraint that the LRP will only translate within the same address family.

I would be OK with adding an API description to the field `RedirectFrontend` in `CiliumLocalRedirectPolicySpec` that states that the `RedirectFrontend` will translate traffic when the address family in the `RedirectFrontend` matches the address family in the `RedirectBackend`. Plus a comment that if the service has an address in a family that does not match the backend family, the behavior of translation for that service on that family is undefined. (Or "implementation defined"). We could then have a subsequent example comment describing that for a service matcher, services with IPv4 addresses will be translated if the backend specifies an IPv4 address.

(I would not go so far as to describe what the behavior is if the address family differs. At most I would state that there is no guarantee what will happen. We know that for now it's a no-op but I don't know if we want to encourage people to write against the API with that assumption)

Point of this exercise being that doesn't force us to commit to a specific behavior for these ambiguous policies, and technically in future we could extend the API to give clarity to this case. But we only need to do that if we have a specific use case.

liyihuang replied:
in our LRP implementation, we pick all frontend and backend IPs and  filter the backend ip familiy based on the frontend ip familiy. Here is the logic
https://github.com/cilium/cilium/blob/ae1e694372cc35e2bc585dc6726cc3539b1a652c/pkg/loadbalancer/writer/writer.go#L434-L441


> I would be OK with adding an API description to the field RedirectFrontend in CiliumLocalRedirectPolicySpec that states that the RedirectFrontend will translate traffic when the address family in the RedirectFrontend matches the address family in the RedirectBackend. Plus a comment that if the service has an address in a family that does not match the backend family, the behavior of translation for that service on that family is undefined. (Or "implementation defined"). We could then have a subsequent example comment describing that for a service matcher, services with IPv4 addresses will be translated if the backend specifies an IPv4 address.
> 
> (I would not go so far as to describe what the behavior is if the address family differs. At most I would state that there is no guarantee what will happen. We know that for now it's a no-op but I don't know if we want to encourage people to write against the API with that assumption)


I personally think this is more confusing for people who read it since we dont have any problem for dual stack with general LRP but only have the problem when we try to override with this `OverrideIP` or `ToIP` field if we introduce it as a string. Shouldn't we just add more docs for `ToIP` field?

liyihuang replied:
let's say we have dual stack service `svc-test` cluster IP `169.254.169.254/2001::1`
with this new field, you can write front end  `169.254.169.254`  override to  `1.1.1.1` , `2001::1`  override to `2002::2 `with 2 separate LRP rules

since the field is a string, you can't write  front end `svc-test`  by using label match to override the dual stack backend to be `1.1.1.1` and `2002::2`.

Hope its a bit clear

joestringer replied:
My thinking here is that regardless of the current implementation, we should describe in the full API what the intention is regarding cross-family translation. If we never want to perform NAT46/NAT64 with this API, we can commit to that and describe that clearly in the API. If we ever might want that, we could document cross-family translation as undefined or implementation defined. That said, if we currently do not do this then maybe it's better to commit to not performing cross-family translation in this version. Then if we ever want to do that in future, we would introduce a new major API version for the CRD to support that.

I don't think this is an API constraint we should describe in this one nested field, I think this is something that the overall API should be consistent about. Making one nested field in the backends have a specific behavior can be surprising and confusing to users. I understand that in this change it is only adding one field, but this one field has raised this question overall about how the API works, so I think this is the right time to clarify that the API is designed around having an object for each address family.

joestringer replied:
From what I understand, the current proposal to only translate within the same address family is actually consistent with the current implementation for other field types, so that resolves the original question from this thread - how do you do it? You create two objects, one per address family.

We can separately look at the whole API and figure out the right level to document that this is the behavior. I think that it's probably within the spec as part of the frontend/backend objects, because this semantic applies for all the frontend/backend combinations. My earlier suggestion was to state that cross-family is undefined/implementation-defined. However, now I'm leaning more towards stating explicitly that this resource does not perform cross-family translation, ie no NAT46/NAT64.

liyihuang replied:
oh, I didn't think about this from the whole API perspective. Overall your statement makes sense and really appreciate the feedback. I will include some documents in the API to describe the current limitation.


> let's say we have dual stack service svc-test cluster IP 169.254.169.254/2001::1
with this new field, you can write front end 169.254.169.254 override to 1.1.1.1 , 2001::1 override to 2002::2 with 2 separate LRP rules

> since the field is a string, you can't write front end svc-test by using label match to override the dual stack backend to be 1.1.1.1 and 2002::2.

I think twice about this and feel it's doable.  With 2 LRP polices using the labels matching the and use ipv4/ipv6 address in this new field. with the exsiting filter logic, that should just work


I will test locally and update the PR based on the feedback if there is no any other concern.

liyihuang committed:
pkg/lrp: clarify address family matching for LRP

Clarify that RedirectFrontend only translates traffic when address
families match.

Signed-off-by: Liyi Huang <liyi.huang@isovalent.com>

ysksuzuki reviewed (changes requested):
Sorry for the late comment, but I have some concerns about this change from a security and semantics perspective.

`redirectBackend` is intentionally constrained to pods living in the same namespace as the `CiliumLocalRedirectPolicy`, as stated in [the LRP documentation](https://docs.cilium.io/en/stable/network/kubernetes/local-redirect-policy):

> The namespace of backend pod(s) need to match with that of the policy.


By allowing `redirectBackend.toIP` to specify an arbitrary IP address, it seems possible to redirect traffic to endpoints that are not pods in the policy's namespace (for example, Service ClusterIPs or endpoints in other namespaces). This appears to break the original namespace-scoping semantics of LRP backends.

Related to this, it's not clear how `toIP` is verified against the pods selected by `redirectBackend.localEndpointSelector`. Without some form of validation that the specified IP actually belongs to the selected backend pods, `localEndpointSelector` becomes largely meaningless: a user could create an arbitrary dummy pod to satisfy the selector and then override the backend IP via `toIP` to redirect traffic to any destination.

In addition, while CLRP is a namespaced resource, the use of addressMatcher already makes it effectively cluster-wide, since it can hijack traffic to a specific address from all workloads regardless of namespace. Allowing an arbitrary backend address on top of that further increases the blast radius and makes CLRP significantly more invasive.

Given this, I'm concerned that this change effectively turns CLRP into something very close to a generic cluster-wide DNAT mechanism, which feels too powerful compared to the original intent and scoping of LRP. Some additional constraints or guardrails (for example, tighter validation on toIP, or clearer scoping restrictions) may be necessary to avoid breaking the intended security and isolation model.

```yaml
apiVersion: cilium.io/v2
kind: CiliumLocalRedirectPolicy
metadata:
  name: gke-metadata-server-redirect-80
  namespace: my-namespace
spec:
  redirectFrontend:
    addressMatcher:
      ip: "169.254.169.254"
      toPorts:
      - port: "80"
        protocol: "TCP"
  redirectBackend:
    toIP: 10.96.0.10 # Can we specify an endpoint in another namespace? for exmaple, kube-dns cluster IP in kube-system
    localEndpointSelector:
      matchLabels:
        k8s-app: "myapp"
    toPorts:
    - port: "988"
      protocol: "TCP"
```

liyihuang committed:
pkg/lrp: introduce the 	lrp-to-ip-range

Introducing a new flag lrp-to-ip-range in LRP to limit  the range users can use when using `toIP` field.

Signed-off-by: Liyi Huang <liyi.huang@isovalent.com>

liyihuang committed:
helm: Add toIPRange configuration

Add a new Helm value `localRedirectPolicies.toIPRange` to configure the allowed IP range for the ToIP field in Local Redirect Policies. The configuration maps to the `lrp-to-ip-range` flag in the Cilium ConfigMap.

Signed-off-by: Liyi Huang <liyi.huang@isovalent.com>
